crawler.c

```diff
@@ -624,12 +681,25 @@ int lru_crawler_start(uint8_t *ids, uint32_t remaining,
         }
     }
 
-    /* we allow the autocrawler to restart sub-LRU's before completion */
-    for (int sid = POWER_SMALLEST; sid < POWER_LARGEST; sid++) {
-        if (ids[sid])
-            starts += do_lru_crawler_start(sid, remaining);
+    if (ids == NULL) {
+        /* NULL ids means to walk the hash table instead. */
+        starts = 1;
+        /* FIXME: hack to signal hash mode to the crawler thread.
+         * Something more clear would be nice.
+         */
+        crawler_count = -1;
```

Comment:
Would setting this to '-1' affect any pending crawl requests?

For example: Could crawler_count be something greater than 0 when we set it to -1 here? If it could, then that would affect the crawlers yet to run (i.e. the ones linked to 'heads' and 'tails' in item.c)

Reply:
normal crawler shouldn't be able to start if it's already running. EXPIRE is special cased to allow restarts while it's already running. I might want to specifically test for this.

Reply:
Trying to think of an example:

If one client sends "lru_crawler metadump all", and while that's just starting to be serviced (meaning crawler_count is definitely > 0);
another client could send in "lru_crawler metadump hash", which takes the 'lru_crawler_lock' when the item crawler thread is sleeping (because of 'crawls_persleep'), and set crawler_count = -1.

When the item crawler thread wakes up again, it would send the crawler_count to a negative value and all book keeping from then on is incorrect.

Reply:
need to do a wider refactor to make expire work, and to keep the interface generic for modularization; when doing that I'll definitely need a different signal that crawler_count=-1.

Reply:
look at the is_running tests in lru_crawler_start()

Reply:
Ah I missed that. You're right, it only affects EXPIRE. We'd need to still fix that to make sure nothing breaks during regular LRU maintenance , right?

Reply:
shouldn't have to; there's another rule ensuring only one active module at a time. if you want to start metadump it has to finish and pause expire.. there's some cruddy code in there that blocks expire from restarting if you're issuing manual commands and they're bouncing.

Reply:
I'm not really thrilled with the way that ended up but I didn't want to invest a ton of time rewiring it to be a stack/queue at the time, so I ended up just putting the timed block code in to make it work.

Reply:
(it also special cases so hash is metadump only, as I have a few other things to refactor before expire will work with a hash walk :/)

Reply:
I think I'm missing something about expire and hash walk conflicting with each other. Am I right in understanding that a "lru_crawler metadump hash" would not break an already running AUTOEXPIRE, but stop it from running to completion, potentially leaving expired items around for longer than they need to be?

And if yes, would you be submitting further changes to fix that in this patch? Or a follow on patch?

Reply:
It waits for the autoexpire to finish, preventing new sub-LRU autoexpires from firing. only one type of LRU crawler can run at a time so metadump always blocks the expire check; it's always been like that.

The internal autoexpire routine uses a histogram to ensure it's only running when it thinks it can recover a lot of data, and when it does it targets specific LRU's that it think it'll get data from, so typically it probably doesn't run much.

Don't think I'm going to be refactoring the autoexpire stuff as part of this patch series. There's no benefit to having autoexpire run on hash mode so it can be done later. I'm not really liking the code structure so having more time to think about it would help too. If you have more immediate needs I can do a different PR for it sooner than later :)

Reply:
to be more specific, if you try to run metadump while expire is running, the metadump will return an error (crawler already running). internally there's a hack that says "someone is trying to run a manual command; don't start any automated internal commands for another minute), to allow you to re-try the metadump command once every N seconds and it'll eventually guarantee to work.

I'd prefer to just queue them, probably. running multiple types at once is no good for a lot of reasons.

Reply:
Ah this makes it a lot clearer. I re-read the code after your explanation and it looks about right. Other than your code restructuring changes, I think this part looks fine functionally.